vspec/model/constants.py

```diff
@@ -10,127 +10,159 @@
 #
 # Constant Types and Mappings
 #
-
 # noinspection PyPackageRequirements
-from enum import Enum, EnumMeta
-from typing import Sequence, Type, TypeVar
+import re
+import pkg_resources
+from typing import (
+    Sequence, Type, TypeVar, Optional, Dict, Tuple, NamedTuple, Iterator,
+)
+
+import yaml
+
+from vspec.model import CONFIG_PATH, NON_ALPHANUMERIC_WORD
 
 
 T = TypeVar("T")
 
 
-class EnumMetaWithReverseLookup(EnumMeta):
-    """This class extends EnumMeta and adds:
+class VSSConstant(NamedTuple):
+    label: str
+    value: str
+    description: Optional[str] = None
+    domain: Optional[str] = None
+
+
+def dict_to_constant_config(name: str, info: Dict[str, str]) -> Tuple[str, VSSConstant]:
+    label = info['label']
+    label = NON_ALPHANUMERIC_WORD.sub('', label).upper()
+    description = info.get('description', None)
+    domain = info.get('domain', None)
+    return label, VSSConstant(info['label'], name, description, domain)
+
+
+def iterate_config_members(config: Dict[str, Dict[str, str]]) -> Iterator[Tuple[str, VSSConstant]]:
+    for u, v in config.items():
+        yield dict_to_constant_config(u, v)
+
+
+class VSSEnumMeta(type):
+    """This class defines the enumeration behavior for vss:
+     - Access through Class.ATTRIBUTE
      - from_str(str): reverse lookup
      - values(): sequence of values
     """
-    def __new__(typ, *args, **kwargs):
-        cls = super().__new__(typ, *args, **kwargs)
-        if not hasattr(cls, "__reverse_lookup__"):
+
+    def __new__(mcs, cls, bases, classdict):
+        cls = super().__new__(mcs, cls, bases, classdict)
+
+        if not hasattr(cls, '__reverse_lookup__'):
             cls.__reverse_lookup__ = {
-                v.value: v for v in cls.__members__.values()
+                v.value: v for k, v in cls.__config__.items()
             }
-        if not hasattr(cls, "__values__"):
-            cls.__values__ = tuple(v.value for v in cls.__members__.values())
+        if not hasattr(cls, '__values__'):
+            cls.__values__ = set(cls.__reverse_lookup__.keys())
+
         return cls
 
+    def __getattr__(cls, key: str) -> str:
+        try:
+            return cls.__config__[key]
+        except KeyError as e:
+            raise AttributeError(
+                f"type object '{cls.__name__}' has no attribute '{key}'"
+            ) from e
+
+    def from_str(cls: Type[T], value: str) -> T:
+        return cls.__reverse_lookup__[value]
+
+    def values(cls: Type[T]) -> Sequence[str]:
+        return cls.__values__
+
+
+class VSSEnumMetaConfigurable(VSSEnumMeta):
+    def __new__(mcs, cls, bases, classdict):
+        cls = super().__new__(mcs, cls, bases, classdict)
+
+        with open(CONFIG_PATH, mode='r') as config_file:
```

Comment:
this should be done elsewhere, the new class should only receive `__config__`, such as:

```py
# constants.py

def load_default_units_config() -> Dict[str, VSSConstant]:
    # remove CONFIG_PATH from __init__.py,
    config_file = pkg_resources.resource_stream('vspec', 'config.yaml') # not sure if you can use "with", if you can, just use that
    yaml_config = yaml.safe_load(config_file)
    return yaml_config.get('units', {})

class Unit(metaclass=VSSEnumMetaConfigurable):
    __config__ = load_default_units_config()

# in the application that may allow extensions, ex with argparse:
ap.add_argument('--config', type=argparse.FileType('r'))
args = ap.parse()

yaml_config = yaml.safe_load(args.config) # or the app loads as json, or whatever, we just care about a dict
Unit.add_config(yaml_config.get('units', {}))
```

IF the `config.yaml` bundled in vspec is to be shared by multiple readers, you can save some work with a singleton doing "on page fault" read:

_global_config = None

def load_config() -> Dict[str, Any]:
    if _global_config is None:
        config_file = pkg_resources.resource_stream('vspec', 'config.yaml') # not sure if you can use "with", if you can, just use that
        _global_config = yaml.safe_load(config_file)
    return _global_config

def load_default_units_config() -> Dict[str, VSSConstant]:
    return load_config().get('units', {})

def load_default_some_other_stuff_config() -> Dict[str, VSSConstant]:
    return load_config().get('some-other-stuff', {})
```

Reply:
Done